### DISCUSSION TITLE
Why the specification doesn't include a synchronous gauge?

### DISCUSSION THREAD

**Discussion by gabrielgiussi** [open-telemetry/opentelemetry-specification#2882] [Q&A] [answered]
I have a use case where using a synchronous gauge will make things simpler, although I was able to implement a solution based on the callbacks I would like to understand if there are technical reasons to not implement a sync api for the gauge or do you understand it is conceptually wrong for some reason I can't see.

My use case is reporting metrics from datomic peer library which is also [callback based](https://docs.datomic.com/on-prem/operation/monitoring.html#custom), so I can't ask for this metrics to datomic whenever the `MetricReader` ends up the callbacks from the gauge instruments.
To overcome this I just stored the last reported metric from datomic in an `AtomicReference` and I read from it when the gauge callback is invoked.
A synchronous API would allow me to call the `ObservableDoubleMeasurement#record` method directly in the datomic callback.

**Comment by jack-berg** [comment, marked as answer]
Synchronous gauge is a rather popular [request](https://github.com/open-telemetry/opentelemetry-specification/issues/2318), which I think will likely be added to the spec in the future. The workaround you've described is how I would solve this in the interim - obviously not ideal.

**Comment by gabrielgiussi** [comment]
@jack-berg perhaps you can help me with a question that raises with this solution that won't appear with the sync api...
After calling record should I clean the atomic reference or not? If I don't clean it and the gauge callback happens before the datomic callback then I would be reporting the same value twice, on the other hand if I clean it and the same occurs then I would have a gap, but I can't understand how this would affect the reported metric so I can't make a justified decision.

**Reply by jack-berg** [reply]
OpenTelemetry asynchronous instruments (i.e. those with callbacks) are responsible for managing their own streams. That means that for a particular collection, if a callback doesn't record a value for a set of attributes, then no metric point will be collected for that set of attributes. 

This is in contrast to synchronous instruments, which will continue to collect values for attributes which have no recorded values if the instrument aggregation temporality is cumulative. So synchronous delta instrument will only report the diff since the last collection, but synchronous cumulative instruments will report the cumulative set of recordings. 

For asynchronous instruments, the temporality of the instrument doesn't matter (and furthermore gauges have no temporality), so the callback itself is responsible for determining the set of data points (each data point is a value and a distinct set of attributes) returned for a given collection.
